### ISSUE TITLE
Checked nullable variable still resulting in a null error when used inside an arrow function

### ISSUE THREAD

**Issue by SayakMukhopadhyay** [microsoft/TypeScript#46853]
# Bug Report

I am not sure how to explain this but when I declare a nullable variable and then check it with the non null guard, Typescript still shows me that the variable could be null when used inside an arrow function. When using the same variable outside the arrow function, TS doesn't complain. Neither does it complain if I assign a new variable to the nullable variable and use it inside the arrow function.
### 🕗 Version & Regression Information

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwObArBQCGMwAQngM4BKwksAFJimEjAFxwpIC2AI2IBKbgDcImACZwA3gCg4SuABtg8AO4ALCCW4ARAPICAVsAxwAPjyQqVAbkXLtuuAF44UdUigoAsuo6UixsHMJOSpgIcEwAhC4kwnIRykowWlAQGjzA2QCiUJlQTMKOqQC+KWg41PBoxmYY1O5wANoKqalkMFAAjNwA5DjAAwA0KV0wPQBM3L0p5QC6KeS1AekQwfWm5jDUrQAMi6NwCQB0UgCCU1DhVQ271Gd8JGBMTGhJbgB8E8qrMHWQQ+J3OVxu4VSpXkVRqdS05gA1i0EjDUttGntnq93p93L9Ov9gGtApsQXA0Ai0IiLtcepDlNDKvJkOgsDg4ACgWSwCRSHxuABhB4YE6sdhcOC1KCsPCiKU9WXJVJeGA+XC8-lnbp9OAAajg4o48mZMAAnmBgHBhTsLB4Osodf0FTKUHgyo6brMbIJiI5yo55ObLXAjLb4PaUuCetxpbL-TDWRhsLhVerucEjZLeL7bgYRfBrLw7MrlFEYlm4ABSODTdxuDy9JIOwlp3ylwmpaNQQYAgZ-JSVCopNu4YsqE1AA)

### 💻 Code

```ts
export function generateBgsReport(input: number): void {
    let whoa: DObject | null;
    whoa = returnMethod(input)
    if (!whoa) {
        throw new Error();
    }
    const cObjects = [{
        attr1: 'one',
        attr2: 1
    }]
    testMethod(cObjects[0], whoa.dAttr)

    cObjects.map((c) =>
        testMethod(c, whoa.dAttr)
    );

    const check = whoa

    cObjects.map((c) =>
        testMethod(c, check.dAttr)
    );
}

function testMethod(param: CObject, input: string): string {
    return param.attr1 + input
}

type CObject = {
    attr1: string;
    attr2: number;
};

type DObject = {
    dAttr: string;
}

function returnMethod(input: number): DObject | null {
    if (input % 2 === 1) {
        return {
            dAttr: 'test'
        }
    }
    return null
}
```

### 🙁 Actual behavior
A checked nullable variable used inside a arrow function is showing that it still could be null

### 🙂 Expected behavior

After I check a nullable variable outside a arrow function, it should recognize this check even inside an arrow function

**Comment by RyanCavanaugh** [comment]
See #11498 or its many inbound duplicates

**Comment by fatcerberus** [comment]
Also see #9998.
